### PR TITLE
chore: revoke badges

### LINKED ISSUE #1184: Add revokeBadge and test impact on airdrop

### PR DESCRIPTION
**Pull request by 1emu**
Closes #1184 

image

### CODE REVIEW

File: src/services/BadgesService.ts [hunk 1]
```diff
@@ -156,4 +165,52 @@ export class BadgesService {
...
     }
   }
+
+  static async revokeBadge(
+    badgeCid: string,
+    addresses: string[],
+    reason = OtterspaceRevokeReason.TenureEnded
+  ): Promise<RevocationResult[]> {
+    console.log('badgeCid', badgeCid)
+    console.log('addresses', addresses)
```
Comment (andyesp) [inline src/services/BadgesService.ts:175]: console.logs

File: src/entities/Badges/utils.ts [hunk 2]
```diff
@@ -12,16 +14,14 @@ function checksumAddresses(addresses: string[]): string[] {
...
-  formattedRecipients: string[],
-  ipfsAddress: string,
-  provider: JsonRpcProvider
+  estimateFunction: (...args: any[]) => Promise<any>
 ): Promise<GasConfig> {
-  const gasLimit = await contract.estimateGas.airdrop(formattedRecipients, ipfsAddress)
+  const provider = RpcService.getPolygonProvider()
+  console.log('estimating gas limit')
+  const gasLimit = await estimateFunction()
+  console.log('gasLimit', gasLimit)
```
Comment (andyesp) [inline src/entities/Badges/utils.ts:24]: console.logs

File: src/back/routes/badges.ts [hunk 3]
```diff
@@ -35,5 +36,39 @@ async function airdropBadges(req: WithAuth): Promise<AirdropOutcome> {
     validateAddress(address)
   })
 
-  return await BadgesService.giveBadgeToUsers(badgeSpecCId, recipients)
+  if (!badgeSpecCid || badgeSpecCid.length === 0) {
+    throw new RequestError('Invalid Badge Spec Cid', RequestError.Unauthorized)
```
Comment (andyesp) [inline src/back/routes/badges.ts:40]: i'd send a bad request error here

File: src/back/routes/badges.ts [hunk 4]
```diff
@@ -35,5 +36,39 @@ async function airdropBadges(req: WithAuth): Promise<AirdropOutcome> {
...
+  if (!isDebugAddress(user)) {
+    throw new RequestError('Invalid user', RequestError.Unauthorized)
+  }
+
+  recipients.map((address) => {
+    validateAddress(address)
+  })
+
+  if (!badgeSpecCid || badgeSpecCid.length === 0) {
+    throw new RequestError('Invalid Badge Spec Cid', RequestError.Unauthorized)
```
Comment (andyesp) [inline src/back/routes/badges.ts:60]: here too

File: src/back/routes/badges.ts [hunk 5]
```diff
@@ -35,5 +36,39 @@ async function airdropBadges(req: WithAuth): Promise<AirdropOutcome> {
...
+  }
+
+  const validatedReason = reason
+    ? toOtterspaceRevokeReason(reason, (reason) => {
+        throw new RequestError(`Invalid revoke reason ${reason}`, RequestError.Unauthorized)
```
Comment (andyesp) [inline src/back/routes/badges.ts:65]: and here

File: src/services/BadgesService.ts [hunk 6]
```diff
@@ -156,4 +165,50 @@ export class BadgesService {
...
+        try {
+          await revokeBadge(trimmedId, Number(reason))
+          return {
+            status: RevocationStatus.Success,
+            address: badgeOwnership.address,
+            badgeId: trimmedId,
+          }
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore
+        } catch (error: any) {
```
Comment (andyesp) [inline src/services/BadgesService.ts:201]: we should find a way to type these errors 🥲

**Review by andyesp — APPROVED** [review]

File: src/entities/Badges/utils.ts [hunk 7]
```diff
@@ -1,9 +1,12 @@
...
 import { abi as BadgesAbi } from '@otterspace-xyz/contracts/out/Badges.sol/Badges.json'
+import logger from 'decentraland-gatsby/dist/entities/Development/logger'
 import { ethers } from 'ethers'
 
 import RpcService from '../../services/RpcService'
+import { OTTERSPACE_DAO_RAFT_ID } from '../Snapshot/constants'
+
+import { GAS_MULTIPLIER, GasConfig } from './types'
 
 const RAFT_OWNER_PK = process.env.RAFT_OWNER_PK || ''
```
Comment (ncomerci) [inline src/entities/Badges/utils.ts:11]: Are we sure that this pk cannot be seen from the front?

**Review by ncomerci — APPROVED** [review]

File: src/entities/Badges/utils.ts (hunk 7 above)
Comment (1emu) [reply src/entities/Badges/utils.ts:11]: I think that for that to happen the PK should be in `prd.json`

**Pull request merged**